File: scripts/select-expression-default.mjs
import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) continue;
    const key = item.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    out[key] = next;
    i += 1;
  }
  return out;
}

function toNum(v, fallback = 0) {
  const n = Number(v);
  return Number.isFinite(n) ? n : fallback;
}

function runSuite({ dir, outdir, style }) {
  const args = [
    "scripts/eval-expression-suite.mjs",
    "--dir",
    dir,
    "--outdir",
    outdir,
    "--style",
    style
  ];
  const res = spawnSync(process.execPath, args, {
    cwd: process.cwd(),
    stdio: "pipe",
    encoding: "utf8"
  });
  if (res.status !== 0 && res.status !== 2) {
    throw new Error(`suite_failed style=${style}\n${res.stderr || res.stdout}`);
  }
  const summaryPath = path.join(outdir, "summary.json");
  if (!fs.existsSync(summaryPath)) {
    throw new Error(`summary_not_found style=${style} path=${summaryPath}`);
  }
  return JSON.parse(fs.readFileSync(summaryPath, "utf8"));
}

function scoreSummary(summary) {
  const m = summary?.metrics || {};
  const gate = summary?.gate?.status === "pass";
  const auto = toNum(m.avgAutoTransitionDelta, 99);
  const over = toNum(m.avgOverrideTransitionDelta, 99);
  const changed = toNum(m.avgChangedIntentSegments, 99);
  const coverage = toNum(m.avgOverrideIntentSegments, 0);

  let score = over * 0.55 + auto * 0.35 + changed * 0.1;
  if (!gate) score += 3.5;
  if (coverage < 0.8) score += 2;
  return Number(score.toFixed(6));
}

function ensureDir(dirPath) {
  if (!fs.existsSync(dirPath)) fs.mkdirSync(dirPath, { recursive: true });
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const testsDir = path.resolve(String(args.dir || "tests/expressions"));
  const styleList = String(args.styles || "tegang,natural,sinematik,narator_tegas,melankolis")
    .split(",")
    .map((s) => String(s || "").trim())
    .filter(Boolean);
  const baseOutDir = path.resolve(String(args.outdir || "outputs/eval_select"));
  const cfgPath = path.resolve(String(args.config || "config/expression/defaults.json"));

  ensureDir(baseOutDir);
  ensureDir(path.dirname(cfgPath));

  const results = [];
  for (const style of styleList) {
    const styleOut = path.join(baseOutDir, style);
    ensureDir(styleOut);
    const summary = runSuite({
      dir: testsDir,
      outdir: styleOut,
      style
    });
    const score = scoreSummary(summary);
    results.push({
      style,
      score,
      gate: summary?.gate?.status || "unknown",
      metrics: summary?.metrics || {},
      summaryFile: path.relative(process.cwd(), path.join(styleOut, "summary.json"))
    });
    console.log(`style_eval style=${style} score=${score} gate=${summary?.gate?.status || "unknown"}`);
  }

  results.sort((a, b) => a.score - b.score);
  const best = results[0];
  if (!best) {
    throw new Error("no_results");
  }

  const payload = {
    updatedAt: new Date().toISOString(),
    source: {
      testsDir: path.relative(process.cwd(), testsDir),
      candidateStyles: styleList
    },
    selected: {
      style: best.style,
      score: best.score,
      gate: best.gate
    },
    ranking: results
  };
  fs.writeFileSync(cfgPath, JSON.stringify(payload, null, 2), "utf8");
  console.log(`expression_default_selected style=${best.style} config=${path.relative(process.cwd(), cfgPath)}`);
}

main();

File: scripts_py/select_expression_default.py
import json
import os
from pathlib import Path
from datetime import datetime, timezone

from common_pipeline import ensure_dir, parse_args, run_cmd


def to_num(v, fallback=0.0):
    try:
        n = float(v)
        if n == n and n not in (float("inf"), float("-inf")):
            return n
    except Exception:
        pass
    return fallback


def score_summary(summary):
    m = summary.get("metrics", {})
    gate = summary.get("gate", {}).get("status") == "pass"
    auto = to_num(m.get("avgAutoTransitionDelta"), 99)
    over = to_num(m.get("avgOverrideTransitionDelta"), 99)
    changed = to_num(m.get("avgChangedIntentSegments"), 99)
    coverage = to_num(m.get("avgOverrideIntentSegments"), 0)
    score = over * 0.55 + auto * 0.35 + changed * 0.1
    if not gate:
        score += 3.5
    if coverage < 0.8:
        score += 2
    return round(score, 6)


def run_suite(tests_dir, out_dir, style):
    cmd = [
        "python",
        "scripts_py/eval_expression_suite.py",
        "--dir",
        str(tests_dir),
        "--outdir",
        str(out_dir),
        "--style",
        str(style),
    ]
    run_cmd(cmd, allowed_exit_codes=(0, 2))
    summary_path = Path(out_dir) / "summary.json"
    if not summary_path.exists():
        raise RuntimeError(f"summary_not_found style={style} path={summary_path}")
    with open(summary_path, "r", encoding="utf-8") as f:
        return json.load(f)


def main():
    args = parse_args(os.sys.argv[1:])
    tests_dir = Path(str(args.get("dir") or "tests/expressions")).resolve()
    styles = [s.strip() for s in str(args.get("styles") or "tegang,natural,sinematik,narator_tegas,melankolis").split(",") if s.strip()]
    base_out = Path(str(args.get("outdir") or "outputs/eval_select")).resolve()
    config = Path(str(args.get("config") or "config/expression/defaults.json")).resolve()
    ensure_dir(str(base_out))
    ensure_dir(str(config.parent))

    results = []
    for style in styles:
        style_out = base_out / style
        ensure_dir(str(style_out))
        summary = run_suite(tests_dir, style_out, style)
        score = score_summary(summary)
        gate = summary.get("gate", {}).get("status", "unknown")
        print(f"style_eval style={style} score={score} gate={gate}")
        results.append(
            {
                "style": style,
                "score": score,
                "gate": gate,
                "metrics": summary.get("metrics", {}),
                "summaryFile": str((style_out / "summary.json").relative_to(Path.cwd())).replace("\\", "/"),
            }
        )
    results.sort(key=lambda x: x.get("score", 9999))
    best = results[0]
    payload = {
        "updatedAt": datetime.now(timezone.utc).isoformat(),
        "source": {
            "testsDir": str(tests_dir.relative_to(Path.cwd())).replace("\\", "/"),
            "candidateStyles": styles,
        },
        "selected": {"style": best["style"], "score": best["score"], "gate": best["gate"]},
        "ranking": results,
    }
    with open(config, "w", encoding="utf-8") as f:
        json.dump(payload, f, indent=2, ensure_ascii=False)
    print(f"expression_default_selected style={best['style']} config={str(config.relative_to(Path.cwd())).replace('\\', '/')}")


if __name__ == "__main__":
    main()
